### DISCUSSION TITLE
ELN File Import

### DISCUSSION THREAD

**Discussion by nicobrandt** [TheELNConsortium/TheELNFileFormat#27] [General]
I'm not sure how far everyone is yet with importing (let alone exporting) ELN files, so I thought I would share some code @jmanideep and me wrote for importing ELN files to Kadi4Mat as a reference: https://gitlab.com/iam-cms/kadi-apy/-/blob/develop/kadi_apy/lib/imports.py#L27

The import is currently done via the Kadi4Mat REST API using our Python wrapper library, although we might built it into the GUI/API directly eventually, at least for smaller crates. It is currently also limited to the very basic metadata that every ELN should probably be able to export. Some Kadi4Mat-specific things might be added as well later on based on some of the file contents, since in our case these can include exported JSON, RDF, etc.

The idea why I'm sharing this is that it might help others in their implementation or even facilitate the creation of common libraries in the future, as there are quite a few things and edge-cases that have to be considered, even when just focusing on the basics. You will notice this when looking at the excessive error handling we added in the code. Some of these aspects also came up in discussions with @NicolasCARPi (originally in https://github.com/TheELNConsortium/TheELNFileFormat/pull/26), such as handling of certain file names. This is less of an issue when just running the import locally instead of on a server, but should still be taken into account.

Let us know what you think :)

**Comment by NicolasCARPi** [comment]
> even facilitate the creation of common libraries in the future

Yeah, I was thinking this too. Basically a lib so you feed it a path to a .eln and it gives you a python object with the json interpreted or something.

**Reply by nicobrandt** [reply]
Right, something along these lines. Concerning Python, it is worth mentioning that https://github.com/ResearchObject/ro-crate-py/ already exists. It didn't really fit our needs for the export, but we should check again whether it's worth using for the import.
